Fix share permission checks

PVince81 opened the pull request:
Fixes https://github.com/owncloud/core/issues/13193

@icewind1991 @schiesbn 
- [x] add unit test

Commit:
Fix shared storage permission checks

icewind1991 commented:
It should still check for updatable permissions

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/247d1a59-1636-4ac9-8c3e-9b8cecd40def) was created.

PVince81 commented:
I added more unit tests for permissions, it confirmed what you meant.
Also, it fixes additional permission issue.

I haven't tested this extensively yet, need to do that later.
Setting to 8.0.1

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/247d1a59-1636-4ac9-8c3e-9b8cecd40def) completed: **5 new issues, 5 updated code elements**

DeepDiver1975 commented:
pr is against master -> milestone 8.1

PVince81 commented:
@owncloud-bot retest this please

PVince81 commented:
Trashbin test side-effect, nice... Running those tests alone work.

- [ ] fix unit tests

PVince81 commented:
Commenting out the rename test in the test removes the weird side effect: https://github.com/owncloud/core/pull/13802/files#diff-c4b4f36fc675d709a1256508ee31c76fR221

Needs further research...

PVince81 commented:
Okay, go it. The "pre_rename" hook disables the trash and "post_rename" reenables it.
But here since rename failed, the "post_rename" is never called... and I'm not sure we want to always call "post_rename" on error. Need to find a better way.

PVince81 commented:
@owncloud-bot retest this please

PVince81 commented:
Ah yes, these were the tests:
```
There were 3 failures:

1) OCA\Files_trashbin\Tests\Storage\Storage::testSingleStorageDelete
Failed asserting that 0 matches expected 1.

/srv/www/htdocs/owncloud/apps/files_trashbin/tests/storage.php:80

2) OCA\Files_trashbin\Tests\Storage\Storage::testCrossStorageDelete
Failed asserting that 0 matches expected 1.

/srv/www/htdocs/owncloud/apps/files_trashbin/tests/storage.php:106

3) OCA\Files_trashbin\Tests\Storage\Storage::testDeleteVersions
Failed asserting that 0 matches expected 1.

/srv/www/htdocs/owncloud/apps/files_trashbin/tests/storage.php:131
```

Commit:
Reenable trashbin after failed rename

After a failed rename, the post_rename hook is not called.
This quickfix makes sure the trashbin storage logic is reenabled
also after a failed rename.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/cccff685-5cc4-4f58-8a54-7f3d0612cc0f) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/cccff685-5cc4-4f58-8a54-7f3d0612cc0f) completed: **4 new issues, 6 updated code elements**

PVince81 commented:
#### Test plan (to be tested with WebDAV and Web UI)
- [ ] create only
- [ ] change only
- [ ] delete only
- [ ] create+change
- [ ] change+delete
- [ ] create+change+delete

PVince81 commented:
(yes, some combinations are a bit ridiculous, please vote for merging of permissions here https://github.com/owncloud/core/issues/12970

PVince81 commented:
@icewind1991 @DeepDiver1975 @schiesbn @nickvergessen please review

PVince81 commented:
and maybe @MorrisJobke :smile:

PVince81 commented:
Rebased onto master (there was an issue related to the obsoleted oc_locks)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ef98fb57-b032-4fd9-9931-9961f41cfad8) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ef98fb57-b032-4fd9-9931-9961f41cfad8) completed: **10 new issues, 6 updated code elements**

PVince81 commented:
Looks like we're not done yet... more inconsistencies:
- [ ] BUG: move file out of "create+delete": web UI works but WebDAV says forbidden
- [ ] BUG: move file out of "delete": web UI works but WebDAV says forbidden

PVince81 commented:
- [ ] BUG (?): move a file "a.txt" into folder "update" (with only update permissions), and "update/a.txt" already exists. Moving a file there is technically overwriting, which is an update, not create. This doesn't work (WebDAV)

PVince81 commented:
Wrote a smashbox test suite to make it easier to test the WebDAV part: https://github.com/owncloud/smashbox/pull/38

PVince81 commented:
This permission thing and interpretations is a real nightmare.

I just found out again that if you try to overwrite a file by moving, Sabre will automatically delete the target (which then requires DELETE permission). But the reason we have the UPDATE permission is to allow people to overwrite files, so that permission should be enough in itself.

It is not possible to add any clean workarounds for this, so I might end up agreeing that "overwrite" needs UPDATE + DELETE permissions... for both upload and rename.

PVince81 commented:
- [ ] BUG: web UI shows rename icon for (update+delete and update) but rename only works if create+update is granted

Commit:
Fix permission checks in Sabre connector

This fixes moving files in and out of shared folders with some exotic
permission combinations.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/2b1c8604-8247-465e-9cac-71ac39784849) was created.

PVince81 commented:
Damn, just compared with master and it looks like so far we always allowed "rename in place" if the update permission was granted.
I'll try to reallow that without breaking the other permissions.

PVince81 commented:
- [ ] fix sabre unit tests for the updated move permission check

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/2b1c8604-8247-465e-9cac-71ac39784849) completed: **6 new issues, 6 updated code elements**

Commit:
Rename must be possible with update-only permission

and this as long as the rename is done within the same folder.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/de8895f3-79c8-4ff6-a3f9-9dcf4400b6eb) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/de8895f3-79c8-4ff6-a3f9-9dcf4400b6eb) completed: **6 new issues, 6 updated code elements**

nickvergessen commented on apps/files_sharing/tests/sharedstorage.php:
```diff
@@ -199,6 +199,154 @@ function testGetPermissions() {
+		$this->assertTrue($result);
+	}
+
+	function testFopenWithDeleteOnlyPermission() {
```
please add `public` next time ;)

Commit:
Fix share permission related unit tests

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/dea7b9b4-c7bc-472f-8fba-328001638f4a) was created.

PVince81 commented:
Fixed the unit tests. This is ready for review.

I'll look into the "overwrite read-only shared file" separately.

PVince81 commented:
@icewind1991 do you know why the permission check logic is duplicated ? One is here https://github.com/owncloud/core/blob/master/lib/private/connector/sabre/objecttree.php#L177 and the other part is in the shared storage.
Maybe we should rename the part in sabre ? (at least I got smashbox tests now to confirm it still works)

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/dea7b9b4-c7bc-472f-8fba-328001638f4a) completed: **2 new issues, 6 updated code elements**

icewind1991 commented:
Looks like the sabre check can go

icewind1991 commented:
looks good :+1:

PVince81 commented:
I'll remove it separately. I tried it locally but ran into complications (aka more work required)

MorrisJobke commented:
Tested and works :+1:

PVince81 commented:
@karlitschek backport ?

karlitschek commented:
please backport

PVince81 commented:
stable8: https://github.com/owncloud/core/pull/15246

karlitschek commented:
thx
